File: admin/app/Actions/Extractors/CleanOcrText.php
<?php

namespace App\Actions\Extractors;

use App\Enums\IdType;

final class CleanOcrText
{
    /**
     * @param  OcrItem[]  $items
     * @param  array<string, IdType>  $labelDictionary
     * @return OcrItem[]
     */
    public function execute(array $items, array $labelDictionary): array
    {
        $labels = array_keys($labelDictionary);
        $result = [];

        foreach ($items as $item) {
            $numberedSplit = $this->trySplitNumberedPrefix($item);

            if ($numberedSplit !== null) {
                foreach ($numberedSplit as $splitItem) {
                    $result[] = $this->autocorrectLabel($splitItem, $labels);
                }

                continue;
            }

            $splitItems = $this->trySplitMergedString($item, $labels);

            foreach ($splitItems as $splitItem) {
                $result[] = $this->autocorrectLabel($splitItem, $labels);
            }
        }

        return $result;
    }

    /**
     * Split items with numbered SIM prefixes (e.g. "1. MELYANA", "2.B LAMPUNG", ".CHANDRA").
     *
     * @return OcrItem[]|null
     */
    private function trySplitNumberedPrefix(OcrItem $item): ?array
    {
        $text = $item->text;

        // Match "N.VALUE" or "N. VALUE" where N is 1-6
        if (preg_match('/^([1-6]\.\s*)(.+)$/s', $text, $matches)) {
            $leftPart = $matches[1];
            $value = trim($matches[2]);
            $label = substr($leftPart, 0, 2);

            if ($value !== '') {
                return $this->buildSplitItems($item, $label, $leftPart, $value);
            }
        }

        // Match ".VALUE" — leading dot is likely a remnant of "1."
        if (preg_match('/^(\.\s*)([A-Z].+)$/', $text, $matches)) {
            $leftPart = $matches[1];
            $value = trim($matches[2]);

            if ($value !== '') {
                return $this->buildSplitItems($item, '1.', $leftPart, $value);
            }
        }

        return null;
    }

    /**
     * @param  string[]  $labels
     * @return OcrItem[]
     */
    private function trySplitMergedString(OcrItem $item, array $labels): array
    {
        // Try colon separators first: "Status Perkawinan: BELUM KAWIN"
        $separators = ['： ', '：', ': ', ':'];

        foreach ($separators as $separator) {
            $pos = strpos($item->text, $separator);

            if ($pos === false || $pos < 2) {
                continue;
            }

            $leftPart = substr($item->text, 0, $pos);
            $rightPart = trim(substr($item->text, $pos + strlen($separator)));

            if ($rightPart === '') {
                continue;
            }

            $matchedLabel = $this->findBestLabelMatch($leftPart, $labels);

            if ($matchedLabel === null) {
                continue;
            }

            return $this->buildSplitItems($item, $matchedLabel, $leftPart, $rightPart);
        }

        // Try space-separated label+value: "Kewarganegaraan WNI", "Status Perkawinan KAWiN"
        $spaceSplit = $this->trySplitOnLabelPrefix($item, $labels);
        if ($spaceSplit !== null) {
            return $spaceSplit;
        }

        // Try label with trailing value (no separator): "Gol.DarahB" → "Gol. Darah" + "B"
        $trailingSplit = $this->trySplitByTrailingValue($item, $labels);
        if ($trailingSplit !== null) {
            return $trailingSplit;
        }

        return [$item];
    }

    /**
     * @param  string[]  $labels
     * @return OcrItem[]|null
     */
    private function trySplitOnLabelPrefix(OcrItem $item, array $labels): ?array
    {
        $bestMatch = null;
        $bestMatchLen = 0;

        foreach ($labels as $label) {
            if (strlen($label) <= 3 || strlen($label) >= strlen($item->text)) {
                continue;
            }

            $prefix = substr($item->text, 0, strlen($label));
            $charAfter = $item->text[strlen($label)] ?? '';

            if ($charAfter !== ' ') {
                continue;
            }

            $distance = levenshtein(mb_strtolower($prefix), mb_strtolower($label));
            $maxLen = max(strlen($prefix), strlen($label));
            $isMatch = $maxLen > 0 && ($distance / $maxLen) < 0.3;

            if ($isMatch && strlen($label) > $bestMatchLen) {
                $bestMatch = $label;
                $bestMatchLen = strlen($label);
            }
        }

        if ($bestMatch === null) {
            return null;
        }

        $rightPart = trim(substr($item->text, $bestMatchLen));

        if ($rightPart === '') {
            return null;
        }

        return $this->buildSplitItems($item, $bestMatch, substr($item->text, 0, $bestMatchLen), $rightPart);
    }

    /**
     * Detect label merged with a short trailing value without any separator.
     * e.g. "Gol.DarahB" → "Gol. Darah" + "B", "Gol.Darah:O" would already
     * be handled by the colon strategy, but "Gol.DarahO" (no colon) would not.
     *
     * @param  string[]  $labels
     * @return OcrItem[]|null
     */
    private function trySplitByTrailingValue(OcrItem $item, array $labels): ?array
    {
        $text = $item->text;
        $textLen = strlen($text);

        if ($textLen < 5) {
            return null;
        }

        // Don't split items that are already a known label
        if ($this->findBestLabelMatch($text, $labels) !== null) {
            return null;
        }

        for ($trim = 1; $trim <= min(3, $textLen - 4); $trim++) {
            $prefix = substr($text, 0, $textLen - $trim);
            $suffix = substr($text, $textLen - $trim);

            $matchedLabel = $this->findBestLabelMatch($prefix, $labels);

            if ($matchedLabel !== null && strlen($matchedLabel) > 3) {
                return $this->buildSplitItems($item, $matchedLabel, $prefix, $suffix);
            }
        }

        return null;
    }

    /**
     * @return OcrItem[]
     */
    private function buildSplitItems(OcrItem $item, string $matchedLabel, string $leftPart, string $rightPart): array
    {
        $totalLen = strlen($item->text);
        $splitRatio = strlen($leftPart) / max($totalLen, 1);

        $leftPoly = $this->splitPolyLeft($item->recPoly, $splitRatio);
        $rightPoly = $this->splitPolyRight($item->recPoly, $splitRatio);

        return [
            new OcrItem($matchedLabel, $leftPoly),
            new OcrItem($rightPart, $rightPoly),
        ];
    }

    /**
     * @param  string[]  $labels
     */
    private function autocorrectLabel(OcrItem $item, array $labels): OcrItem
    {
        $matchedLabel = $this->findBestLabelMatch($item->text, $labels);

        if ($matchedLabel !== null) {
            return $item->withText($matchedLabel);
        }

        return $item;
    }

    /**
     * @param  string[]  $labels
     */
    private function findBestLabelMatch(string $text, array $labels): ?string
    {
        $text = trim($text);

        if ($text === '') {
            return null;
        }

        if (in_array($text, $labels, true)) {
            return $text;
        }

        $bestLabel = null;
        $bestDistance = PHP_INT_MAX;
        $textLower = mb_strtolower($text);

        foreach ($labels as $label) {
            $labelLower = mb_strtolower($label);
            $distance = levenshtein($textLower, $labelLower);

            if ($distance < $bestDistance) {
                $bestDistance = $distance;
                $bestLabel = $label;
            }
        }

        if ($bestLabel === null) {
            return null;
        }

        $maxLen = max(strlen($text), strlen($bestLabel));

        if ($maxLen <= 3) {
            return $bestDistance <= 1 ? $bestLabel : null;
        }

        $threshold = 0.3;

        if ($maxLen > 0 && ($bestDistance / $maxLen) < $threshold) {
            return $bestLabel;
        }

        return null;
    }

    /**
     * @param  array<int, array<int, int>>  $poly
     * @return array<int, array<int, int>>
     */
    private function splitPolyLeft(array $poly, float $ratio): array
    {
        $leftX = $poly[0][0];
        $rightX = $poly[1][0];
        $splitX = (int) ($leftX + ($rightX - $leftX) * $ratio);

        return [
            [$poly[0][0], $poly[0][1]],
            [$splitX, $poly[1][1]],
            [$splitX, $poly[2][1]],
            [$poly[3][0], $poly[3][1]],
        ];
    }

    /**
     * @param  array<int, array<int, int>>  $poly
     * @return array<int, array<int, int>>
     */
    private function splitPolyRight(array $poly, float $ratio): array
    {
        $leftX = $poly[0][0];
        $rightX = $poly[1][0];
        $splitX = (int) ($leftX + ($rightX - $leftX) * $ratio);

        return [
            [$splitX, $poly[0][1]],
            [$poly[1][0], $poly[1][1]],
            [$poly[2][0], $poly[2][1]],
            [$splitX, $poly[3][1]],
        ];
    }
}

File: ocr/extractors/clean_ocr_text.py
from __future__ import annotations

import re

from rapidfuzz.distance import Levenshtein

from .fuzzy_match import passes_threshold
from .id_type import IdType
from .ocr_item import OcrItem


class CleanOcrText:
    def execute(
        self,
        items: list[OcrItem],
        label_dictionary: dict[str, IdType],
    ) -> list[OcrItem]:
        labels = list(label_dictionary.keys())
        result: list[OcrItem] = []

        for item in items:
            numbered_split = self._try_split_numbered_prefix(item)

            if numbered_split is not None:
                for split_item in numbered_split:
                    result.append(self._autocorrect_label(split_item, labels))
                continue

            split_items = self._try_split_merged_string(item, labels)

            for split_item in split_items:
                result.append(self._autocorrect_label(split_item, labels))

        return result

    def _try_split_numbered_prefix(self, item: OcrItem) -> list[OcrItem] | None:
        text = item.text

        m = re.match(r"^([1-6]\.\s*)(.+)$", text, re.DOTALL)
        if m:
            left_part = m.group(1)
            value = m.group(2).strip()
            label = left_part[:2]
            if value:
                return self._build_split_items(item, label, left_part, value)

        m = re.match(r"^(\.\s*)([A-Z].+)$", text)
        if m:
            left_part = m.group(1)
            value = m.group(2).strip()
            if value:
                return self._build_split_items(item, "1.", left_part, value)

        return None

    def _try_split_merged_string(
        self,
        item: OcrItem,
        labels: list[str],
    ) -> list[OcrItem]:
        separators = ["： ", "：", ": ", ":"]

        for separator in separators:
            pos = item.text.find(separator)

            if pos == -1 or pos < 2:
                continue

            left_part = item.text[:pos]
            right_part = item.text[pos + len(separator):].strip()

            if not right_part:
                continue

            matched = self._find_best_label_match(left_part, labels)

            if matched is None:
                continue

            return self._build_split_items(item, matched, left_part, right_part)

        space_split = self._try_split_on_label_prefix(item, labels)
        if space_split is not None:
            return space_split

        trailing_split = self._try_split_by_trailing_value(item, labels)
        if trailing_split is not None:
            return trailing_split

        return [item]

    def _try_split_on_label_prefix(
        self,
        item: OcrItem,
        labels: list[str],
    ) -> list[OcrItem] | None:
        text = item.text
        best_match = None
        best_match_len = 0

        for label in labels:
            label_len = len(label)
            if label_len <= 3 or label_len >= len(text):
                continue

            prefix = text[:label_len]
            char_after = text[label_len:label_len + 1]

            distance = Levenshtein.distance(prefix.lower(), label.lower())

            has_separator = char_after in (" ", ",", ".")
            is_camelcase_boundary = (
                distance == 0 and (char_after.isupper() or char_after.isdigit())
            )

            if not (has_separator or is_camelcase_boundary):
                continue

            if passes_threshold(distance, len(prefix), label_len) and label_len > best_match_len:
                best_match = label
                best_match_len = label_len

        if best_match is None:
            return None

        right_part = text[best_match_len:].strip()

        if not right_part:
            return None

        return self._build_split_items(item, best_match, text[:best_match_len], right_part)

    def _try_split_by_trailing_value(
        self,
        item: OcrItem,
        labels: list[str],
    ) -> list[OcrItem] | None:
        text = item.text
        text_len = len(text)

        if text_len < 5:
            return None

        if self._find_best_label_match(text, labels) is not None:
            return None

        for trim in range(1, min(3, text_len - 4) + 1):
            prefix = text[:text_len - trim]
            suffix = text[text_len - trim:]

            matched = self._find_best_label_match(prefix, labels)

            if matched is not None and len(matched) > 3:
                return self._build_split_items(item, matched, prefix, suffix)

        return None

    def _build_split_items(
        self,
        item: OcrItem,
        matched_label: str,
        left_part: str,
        right_part: str,
    ) -> list[OcrItem]:
        total_len = len(item.text)
        split_ratio = len(left_part) / max(total_len, 1)

        left_poly = self._split_poly_left(item.rec_poly, split_ratio)
        right_poly = self._split_poly_right(item.rec_poly, split_ratio)

        return [
            OcrItem(matched_label, left_poly),
            OcrItem(right_part, right_poly),
        ]

    def _autocorrect_label(self, item: OcrItem, labels: list[str]) -> OcrItem:
        matched = self._find_best_label_match(item.text, labels)

        if matched is not None:
            return item.with_text(matched)

        return item

    def _find_best_label_match(self, text: str, labels: list[str]) -> str | None:
        text = text.strip()

        if not text:
            return None

        if text in labels:
            return text

        best_label = None
        best_distance = float("inf")
        text_lower = text.lower()

        for label in labels:
            distance = Levenshtein.distance(text_lower, label.lower())

            if distance < best_distance:
                best_distance = distance
                best_label = label

        if best_label is None:
            return None

        if passes_threshold(best_distance, len(text), len(best_label)):
            return best_label

        return None

    @staticmethod
    def _split_poly_left(
        poly: list[list[float]],
        ratio: float,
    ) -> list[list[float]]:
        left_x = poly[0][0]
        right_x = poly[1][0]
        split_x = int(left_x + (right_x - left_x) * ratio)

        return [
            [poly[0][0], poly[0][1]],
            [split_x, poly[1][1]],
            [split_x, poly[2][1]],
            [poly[3][0], poly[3][1]],
        ]

    @staticmethod
    def _split_poly_right(
        poly: list[list[float]],
        ratio: float,
    ) -> list[list[float]]:
        left_x = poly[0][0]
        right_x = poly[1][0]
        split_x = int(left_x + (right_x - left_x) * ratio)

        return [
            [split_x, poly[0][1]],
            [poly[1][0], poly[1][1]],
            [poly[2][0], poly[2][1]],
            [split_x, poly[3][1]],
        ]
